indexing-service/src/main/java/io/druid/indexing/common/task/SinglePhaseParallelIndexSupervisorTask.java

```diff
@@ -0,0 +1,765 @@
+              }
+
+              if (!subTaskSpecIterator.hasNext()) {
+                // We have no more subTasks to run
+                if (taskMonitor.getNumRunningTasks() == 0 && taskCompleteEvents.size() == 0) {
+                  stopped = true;
+                  if (taskMonitor.isSucceeded()) {
+                    // Publishing all segments reported so far
+                    publish(toolbox);
+
+                    // Succeeded
+                    state = TaskState.SUCCESS;
+                  } else {
+                    // Failed
+                    throw new ISE(
+                        "Expected for [%d] tasks to succeed, but we got [%d] succeeded tasks and [%d] failed tasks",
+                        taskMonitor.getExpectedNumSucceededTasks(),
+                        taskMonitor.getNumSucceededTasks(),
+                        taskMonitor.getNumFailedTasks()
+                    );
+                  }
+                }
+              } else if (taskMonitor.getNumRunningTasks() < maxNumTasks) {
+                // We have more subTasks to run
+                submitNewTask(taskMonitor, subTaskSpecIterator.next());
+              } else {
+                // We have more subTasks to run, but don't have enough available task slots
+                // do nothing
+              }
+              break;
+            case FAILED:
+              // TaskMonitor already tried everything it can do for failed tasks. We failed.
+              state = TaskState.FAILED;
+              stopped = true;
+              final TaskStatusPlus lastStatus = taskCompleteEvent.getLastStatus();
+              if (lastStatus != null) {
+                log.error("Failed because of the failed sub task[%s]", lastStatus.getId());
+              } else {
+                final SinglePhaseParallelIndexSubTaskSpec spec =
+                    (SinglePhaseParallelIndexSubTaskSpec) taskCompleteEvent.getSpec();
+                log.error(
+                    "Failed to run sub tasks for inputSplits[%s]",
+                    getSplitsIfSplittable(spec.getIngestionSpec().getIOConfig().getFirehoseFactory())
+                );
+              }
+              break;
+            default:
+              throw new ISE("spec[%s] is in an invalid state[%s]", taskCompleteEvent.getSpec().getId(), completeState);
+          }
+        }
+      }
+    }
+    finally {
+      log.info("Cleaning up resources");
+      // Cleanup resources
+      taskCompleteEvents.clear();
+      taskMonitor.stop();
+      chatHandlerProvider.unregister(getId());
+
+      if (state != TaskState.SUCCESS) {
+        log.info(
+            "This task is finished with [%s] state. Killing [%d] remaining subtasks.",
+            state,
+            taskMonitor.getNumRunningTasks()
+        );
+        // if this fails, kill all sub tasks
+        // Note: this doesn't work when this task is killed by users. We need a way for gracefully shutting down tasks
```

Comment:
Thanks for the good suggestion. It sounds working, but I think it would be better for the supervisor task to be responsible for cleaning up its resources including subTasks. I think we can further discuss this in a follow-up issue after merging this PR.